Added byte offsets to rule sets

G0dwin opened the pull request:
I have added some basic logic to the parser to capture byte offsets when scanning for rule sets and added a field to `RuleSet` to retain them. This will allow code coverage and lint utilities (for example) to report where in a file a particular rule was found (see the CSS code coverage utility I'm working on here: [lingua-franca/marmara](https://github.com/lingua-franca/marmara)).

There is still more work that needs to be done, primarily when it comes to differentiating files (from `@import` statement or otherwise), and when urls are expanded.

grosser commented:
idk about this ... kind of neat feature ... but more work/memory for the 99% usecase
only thing striking me as a bit unfortunate is the addition of more optional argument to a few methods ... would be nice to make them keyword args ... but that's not easy without breaking existing usage  ...

@akzhan thoughts ?

akzhan commented:
It's nice contribution, but undone.

We need any use case to be bundled with (may be in examples folder).

akzhan commented:
Also it's not clear to detect line endings by Windows platform.

We may use "read binary mode" to keep line endings of file.

grosser commented:
FYI https://github.com/giakki/uncss

On Thu, Mar 9, 2017 at 12:23 AM, Akzhan Abdulin <notifications@github.com>
wrote:

> Also it's not clear to detect line endings by Windows platform.
>
> We may use "read binary mode" to keep line endings of file.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/premailer/css_parser/pull/85#issuecomment-285285702>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZwR4JguYFEaEOPpY2YBTKWIMbJdbks5rj7cMgaJpZM4MXoM_>
> .
>

G0dwin commented:
@akzhan I can add documentation before merging, however, I'm not seeing a wiki, only the `README.md` file.

@grosser I can refactor so that we only collect offsets if an option is passed to `load_uri!` and try to ensure the former code path remains as close as possible to the original but only add the additional overhead when the option is present.

To get around the additional optional arguments, we could subclass `RuleSet` to something like `FileRuleSet` where the additional code for storing offsets and anything else specific to relating back to the original CSS code could be kept. In order to support import statements and multiple files in general, we will also need to store a filename.

akzhan commented:
@G0dwin ok, we need 

a) a section in README.
b) separate position-aware code.
c) replace CRLF detection with byte-precise code (File.read('rb') etc.).

G0dwin commented:
Alright:
a) I added a section in the [README](https://github.com/lingua-franca/css_parser/blame/master/README.md#L57-L72)
b) I added the option [`:capture_offsets`](https://github.com/lingua-franca/css_parser/blob/master/lib/css_parser/parser.rb#L464) to `load_uri!`, `load_file!`, and `load_string!`, we only look for, process, and store offsets if this flag is set. I also created a subclass of `RuleSet` called [`OffsetAwareRuleSet`](https://github.com/lingua-franca/css_parser/blob/master/lib/css_parser/rule_set.rb#L517-L530). This class takes in additional constructor params and stores the offsets. We only create the subclass if the flag is set.
c) I fixed CRLF detection though fixing the encoding, I didn't change the [`IO.read(...)`](https://github.com/lingua-franca/css_parser/blob/master/lib/css_parser/parser.rb#L462) call. I tested on Windows and Ubuntu machines to ensure the offsets were identical.

In addition, I added a [benchmark rake task](https://github.com/lingua-franca/css_parser/blob/master/Rakefile#L11-L35) (`rake benchmark`) to test the difference between the old code and the new. I don't have an isolated machine to run the tests on but the difference seemed negligible, the last test I ran, parsing `import.css` 50 000 times took `50.7` seconds using the old code and `47.1` seconds with the new. Parsing `screen.css` 5000 times from dialect.ca took `65.9` seconds using the old code, and `63.7` seconds after my change. I would account the decrease to either a random variance, or due to some minor fixes I made to the code base along the way.

I also added a [`filename` attribute](https://github.com/lingua-franca/css_parser/blob/master/lib/css_parser/rule_set.rb#L523) to the `OffsetAwareRuleSet` with which I was able to get imports working with offset capturing. I removed the code I added to tests and isolated offset capturing tests into [one test file](https://github.com/lingua-franca/css_parser/blob/master/test/test_css_parser_offset_capture.rb).

Let me know if there are any other changes that you would like to see before merging, big or small.

Cheers,
godwin

akzhan commented:
LGTM

grosser commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 /pkg/
+/Gemfile.lock
```
please don't https://grosser.it/2015/08/14/check-in-your-gemfile-lock/

G0dwin replied:
This is the `.gitignore`, the change was made in [0ed2f74](https://github.com/premailer/css_parser/commit/0ed2f74e8d6e6ba26c4260aaf4cb4c57159231ec). I'm going to take a closer look though, I'm not as familiar with the way that Github shows diffs, I'm surprised it's showing this and other changes as changes when they currently match the master.

grosser commented on README.md:
```diff
@@ -53,6 +53,22 @@ parser.add_block!(css)
    body { margin: 0 1em; }
+
+# capturing byte offsets within a file
+parser.load_uri!('../style.css', {:base_uri => 'http://example.com/styles/inc/', :capture_offsets => true)
```
prefer 1.9 hash `, base_uri: 'http://'`

G0dwin replied:
I can make this change

akzhan replied:
+

grosser commented on lib/css_parser/parser.rb:
```diff
@@ -289,9 +305,16 @@ def parse_block_into_rule_sets!(block, options = {}) # :nodoc:
         token = matches[0]
 
+        # save the regex offset so that we know where in the file we are
+        offset = Regexp.last_match.offset(0) if options[:capture_offsets]
```
might be worth extracting that into a local variable if this code is used a lot

G0dwin replied:
`options[:capture_offsets]`? Sure.

grosser commented on lib/css_parser/parser.rb:
```diff
@@ -504,11 +577,11 @@ def read_remote_file(uri) # :nodoc:
           res = http.get(uri.request_uri, {'User-Agent' => USER_AGENT, 'Accept-Encoding' => 'gzip'})
           src = res.body
-          charset = fh.respond_to?(:charset) ? fh.charset : 'utf-8'
+          charset = res.respond_to?(:charset) ? res.encoding : 'utf-8'
```
we should get rid of this charset detection and drop 1.8 ... but that's for another PR

grosser commented on lib/css_parser/version.rb:
```diff
@@ -1,3 +1,3 @@
 module CssParser
-  VERSION = "1.4.9".freeze
+  VERSION = "1.4.10".freeze
```
don't change version in the PR

G0dwin replied:
This again seems to be coming from rebasing, I'll take a closer look.

grosser commented on test/test_css_parser_loading.rb:
```diff
@@ -197,10 +197,13 @@ def test_suppressing_circular_reference_exceptions
     end
 
+    uri_regex = Regexp.new(Regexp.escape("#{@uri_base}/no-exist.xyz"))
+    assert_match uri_regex, err.message
```
this might be simpler and produce a more readable error
`err.message.must_include "#{@uri_base}/no-exist.xyz"`

G0dwin replied:
This came from [ed148aa](https://github.com/premailer/css_parser/commit/ed148aab13c142bd3daf8160e82d6644ce29478e#diff-ef9685ca9e71df32025876c2e7ee6535R204)

akzhan replied:
fixed by e2c831e

grosser commented on test/test_css_parser_misc.rb:
```diff
@@ -177,4 +177,9 @@ def test_ruleset_with_braces
+
+  def test_content_with_data
+    rule = RuleSet.new('div', '{content: url(data:image/png;base64,LOTSOFSTUFF)}')
+    assert_match (/image\/png;base64,LOTSOFSTUFF/), rule.to_s
```
`rule.to_s.must_include`

akzhan replied:
fixed by e2c831e also

grosser commented on lib/css_parser/parser.rb:
```diff
@@ -440,20 +499,31 @@ def circular_reference_check(path)
+    # Returns a string.
+    def ignore_pattern(css, regex, options)
+      # if we are capturing file offsets, replace the characters with spaces to retail the original positions
+      return css.gsub(regex) { |m| ' ' * m.length } if options[:capture_offsets]
```
might be easier to read with ` if ... else ...

G0dwin replied:
Sure.

akzhan commented:
Some commits related to rebase (Gemfile.lock, version up), don't worry.

akzhan commented:
Just note that is't released as `1.5.0.pre`.

Other updates should be proposed by other pull requests.
